### DISCUSSION TITLE
development as3?

### DISCUSSION THREAD

**Discussion by skramovskiy** [airsdk/Adobe-Runtime-Support#2833] [General]
The last release of AS was in 2006. Despite the fact that the SDK is constantly evolving, the language itself remains stagnant. Is anyone involved in its development? and should we expect a new version? 17 years is quite a long time.

**Comment by ajwfrost** [comment]
There have been a few minor changes we've made just for convenience - e.g. verbatum strings, nullish coalescing and null-conditional operators. But a lot of people have some other ideas: see https://github.com/airsdk/Adobe-Runtime-Support/discussions/2152

The idea really is that these would be incremental features rather than things that would result in an "ActionScript 4" language definition. AS4 was planned to be done alongside a completely new VM, all of which is a large amount of work .. so we are probably better off making smaller changes from time to time.

The other thing we're looking at doing (and one of the reasons we haven't done much more since those above updates) is to switch the compiler that's used in the AIR SDK over to the one that's used and maintained by Apache Royale. So then we can look at updating the compiler there and adding in the appropriate updates to the JS conversion and to the avmplus runtime, in a more coordinated fashion.

Happy to have other people to get involved in this too, the benefit being that we would have the potential to update the open source Apache Royale compiler along with (when necessary) the open source avmplus code... which of course means that we (Harman) don't end up as the gatekeeper for the innovation!

thanks

**Reply by raresn** [reply]
@ajwfrost is there a list of these minor changes? we kinda missed this. :)

**Reply by ajwfrost** [reply]
The updates were all in one release, so there is a list under this entry:
https://github.com/airsdk/Adobe-Runtime-Support/discussions/1699#discussioncomment-3653584

But I was also just looking at https://airsdk.dev/reference/actionscript/3.0/ and realising that we missed out updating this - both the "What's New" appendix, and the actual details e.g. list of AS3 operators needs updating for "?." etc.

**Reply by shaucky** [reply]
Just out of curiosity and anticipation… Can you introduce the development progress of ActionScript 4.0? For example, the expected percentage, the current possible launch time, and so on… Harman claims to have been planning to develop AS4 for a year or two, right? Also, if AIR provides support for new languages (whether AS4, JS, TS), have you ever planned a compatibility solution with IDEs like Animate? I heard that Adobe once developed an 80% AVM3 project. Will Harman's AVM3 be improved based on this or will it be restarted?

**Reply by ajwfrost** [reply]
Hi - sorry, no to clarify, we're not working on AS4. Adobe had been looking at this, and an updated Flash Player runtime, many years ago, but parked that and eventually published the spec when they decided not to continue. We don't have any code assets or similar from that time.

So rather than a big "AS4" language (which would be incompatible with AS3, although I guess fairly easy to migrate), we are just enhancing the AS3 language. When we add things, they fall into two brackets:
1) compiler-only: where the compiler contains all the extra functionality/logic - for example for verbatim strings - and this then generates ActionScript Byte Code that's compatible with older runtimes
2) compiler-and-runtime: where there need to be additional changes to the runtime, and therefore the ABC spec and avmplus implementation. For example adding generics would most likely be done via this templated mechanism - extending how the 'vector' class is implemented.

Compatibility-wise for IDEs, we'd certainly want to work with BowlerHat (for VSCode) and Moonshine to ensure updates work properly, and we're in touch with the Animate team and exploring ways we could get some updates into that.

Supporting new languages (like TypeScript) is something that would need to be done via a "TypeScript to SWF" type compiler, similar probably to how we currently have a "C++ to SWF" compiler in the Crossbridge toolchain.

**Reply by Dextercitox** [reply]
Personally I think it's a wise decision from Harman. It's better to improve what we have (and in my opinion is quite good overall), than trying to make such a big step with AS4.

In my enterprise we really build large apps with AS3 and the language it's not a problem at all. In fact, it's very easy to learn for any programmer, easy to read and maintain.

**Reply by 2jfw** [reply]
>   The other thing we're looking at doing (and one of the reasons we haven't done much more since those above updates) is to switch the compiler that's used in the AIR SDK over to the one that's used and maintained by Apache Royale. So then we can look at updating the compiler there and adding in the appropriate updates to the JS conversion and to the avmplus runtime, in a more coordinated fashion.

Would this actually enable you to update the flex-specific compiler in order to have these new features available there, too?

**Reply by pnico** [reply]
> compiler-and-runtime: where there need to be additional changes to the runtime, and therefore the ABC spec and avmplus implementation. For example adding generics would most likely be done via this templated mechanism - extending how the 'vector' class is implemented.

I wonder if this could be done (at least initially anyway) like Java does generics where it's just a compile-time check, IDE support etc but the actual emitted bytecode is just ignoring any newly-added generic annotations when it's not `Vector`

**Reply by shaucky** [reply]
How about thinking from another perspective? When the new features of ActionScript 3.0 are updated to a certain stage/step (or likes…), call it ActionScript 4.0. Just like the ralationship in ActionScript 1.0&2.0. (As a way with #2588 , it seems that the negotiation with Adobe failed?)

**Comment by teotigraphix** [comment]
> Supporting new languages (like TypeScript) is something that would need to be done via a "TypeScript to SWF" type compiler, similar probably to how we currently have a "C++ to SWF" compiler in the Crossbridge toolchain.

Heh, TypeScript code compiles SWF from TypeScript, sounds doable! haha
